File: ts/parity/builder_surface_paths.py
"""Machine-independent recording of Builder surfaces that print filesystem paths.

Shared by `generate_builder_lifecycle_golden.py` (function level) and
`generate_builder_command_golden.py` (command level), because both grade the same
surfaces and the rule must not drift between them.

The problem this solves, measured at CV22.DS7.US8 plateau 3:

`story_paths` resolves the roadmap root to an ABSOLUTE path deliberately --
`create_story_directory`'s escape guard is `target.is_relative_to(roadmap_root)`,
which is only sound on absolutes. Two Ariad surfaces then print those paths raw:
`plan_checkpoint` (`story package`, `artifacts`, and the `*_path=` trailer) and
`expand_decision` / `expand_blocked` (`materialized`, `why blocked`). Every other
artifact surface relativizes through `artifact_surfaces._display_path`.

So those regions are machine-dependent, and the obvious repair does not work: the
paths are WRAPPED at 52 or 54 code points by `_card_prefixed` / `_card_wrapped`,
so the absolute prefix's length decides where each chunk boundary falls. A token
substitution matches nothing in
`\u2502 /private/var/folders/5k/q6bjkgn95gzd_qss7_5flyzw0000gp \u2502` -- a truncated prefix
of the root -- and even a successful match would leave machine-dependent
continuation rows.

Two rules, applied at record time:

- `normalize_path_rows` collapses a run of card rows that render an ABSOLUTE path
  into ONE stable token row, preserving the run's first list marker.
- `scrub_message` rewrites absolute paths inside refusal messages, which carry
  them untruncated and can therefore be substituted exactly.

A run is recognized by its FIRST row starting with `/`, and that detail is
load-bearing. The first version asked only whether a row's text was a fragment of
some absolute path, which is also true of every RELATIVIZED row -- `docs/project/…`
is a substring of `/abs/prefix/project/docs/project/…`. It therefore collapsed the
`artifacts_materialized` rows too, erasing content that is machine-independent and
must stay graded byte for byte. Only a row that begins the absolute path can open a
run; relative rows are left exactly as Python printed them.

What remains graded, and why it is enough: the path CONTENT is graded exactly in
each step's structured project-relative fields; the WRAPPING mechanism is graded
exhaustively by `builder-card` (516 rows, including over-long words chunked at 52
and 54); `artifacts_materialized` renders the same paths relativized and is graded
whole; and every other byte of the affected surfaces -- block order, labels,
framing, markers -- is still compared byte for byte.
"""

from __future__ import annotations

import re
from pathlib import Path

PATH_ROW_TOKEN = "<ABSOLUTE PATH>"
OUTSIDE_TOKEN = "<OUTSIDE PROJECT>"

# The list markers `_card_prefixed` can put in front of a wrapped path, plus the
# two-space continuation it uses for a wrapped item's later lines.
_PATH_ROW_MARKERS = ("✓ ", "○ ", "- ", "• ", "✕ ", "✎ ", "↻ ", "  ")
_ABSOLUTE_PATH_RE = re.compile(r"/[^\s,'\"]+")

# Shorter than this, a `/`-leading fragment could still be ordinary content (a card
# could legitimately print `/tmp`), so a run only opens on a long fragment; the
# short final chunk of a wrapped path is captured by the run-continuation clause.
_MIN_STANDALONE_FRAGMENT = 24


def absolute_paths_in(message: str) -> list[str]:
    """Absolute paths embedded in a message, longest first.

    Longest first so replacing a parent directory cannot corrupt a child path that
    shares its prefix.
    """
    found = {match.group(0).rstrip(".") for match in _ABSOLUTE_PATH_RE.finditer(message)}
    return sorted(found, key=len, reverse=True)


def normalize_path_rows(text: str, absolute_paths: list[str]) -> str:
    """Collapse card rows rendering an absolute path into one stable token row."""
    if not absolute_paths:
        return text
    normalized: list[str] = []
    in_run = False
    for line in text.split("\n"):
        fragment, marker = _card_row_content(line)
        if not fragment:
            in_run = False
            normalized.append(line)
            continue
        belongs = any(fragment in path for path in absolute_paths)
        opens_run = (
            fragment.startswith("/") and belongs and len(fragment) >= _MIN_STANDALONE_FRAGMENT
        )
        if opens_run or (in_run and belongs):
            if not in_run:
                normalized.append(f"│ {marker}{PATH_ROW_TOKEN}".ljust(57) + "│")
            in_run = True
            continue
        in_run = False
        normalized.append(line)
    return "\n".join(normalized)


def normalize_trailer_paths(text: str, absolute_paths: list[str], project_root: Path) -> str:
    """Rewrite `key=<absolute path>` trailer lines as project-relative.

    `render_plan_checkpoint` appends `story_package_path=`, `index_artifact_path=`,
    `plan_artifact_path=`, and `test_guide_artifact_path=` AFTER the card, as plain
    lines. Those are unwrapped, so unlike the card rows they can be substituted
    exactly and stay fully graded.
    """
    for absolute in absolute_paths:
        text = text.replace(absolute, project_relative(Path(absolute), project_root))
    return text


def scrub_message(message: str, *, project_root: Path, repo_root: Path | None = None) -> str:
    """Rewrite absolute paths inside a refusal message as relative ones.

    Idempotent on purpose. An earlier version was applied twice to the same message
    -- once at the call site, once while recording -- and the second pass re-matched
    the RELATIVE result (`docs/project/roadmap/…` contains `/project/roadmap/…`) and
    rewrote it to a token, producing `authored package at docs<OUTSIDE PROJECT>`.
    Only candidates under `repo_root` are touched, so a second pass is a no-op and
    an unrelated `/`-prefixed fragment in product text is left alone.
    """
    root = (repo_root or Path.cwd()).resolve()
    project = project_root.resolve()
    for absolute in absolute_paths_in(message):
        if not absolute.startswith(str(root)):
            continue
        candidate = Path(absolute)
        if absolute.startswith(str(project)):
            replacement = project_relative(candidate, project)
        else:
            replacement = f"<REPO>/{candidate.resolve().relative_to(root).as_posix()}"
        message = message.replace(absolute, replacement)
    return message


def project_relative(path: Path, project_root: Path) -> str:
    """Path content in the only form that is machine-independent."""
    resolved = Path(path).resolve()
    root = project_root.resolve()
    if resolved == root:
        return "."
    if not resolved.is_relative_to(root):
        return OUTSIDE_TOKEN
    return resolved.relative_to(root).as_posix()


def _card_row_content(line: str) -> tuple[str, str]:
    """A card row's inner text without its frame and list marker."""
    if not (line.startswith("│ ") and line.endswith("│")):
        return "", ""
    inner = line[2:-1].rstrip()
    for marker in _PATH_ROW_MARKERS:
        if inner.startswith(marker):
            return inner[len(marker) :].strip(), marker
    return inner, ""

File: ts/test/helpers/builderSurfacePaths.ts
// The TypeScript twin of `ts/parity/builder_surface_paths.py`.
//
// Test-only, and it must stay that way: production code renders exactly what
// Python renders, absolute paths included. This is the recording rule the corpus
// was generated under, so the comparison has to apply it to the TypeScript output
// too — otherwise the test compares a normalized golden against an unnormalized
// actual and fails for a reason that has nothing to do with parity.
//
// Why normalization is needed at all: `expand_decision` and `expand_blocked` print
// paths that `storyPaths` resolved to absolute (its confinement guard is only
// sound on absolutes), and `cardPrefixed` / `cardWrapped` then wrap them at 52 or
// 54 code points — so the absolute prefix's length decides where every chunk
// boundary falls, and the rows cannot match across two machines. CR082 is the
// record; the Python module's docstring carries the full rationale and the list of
// what stays graded.
//
// Keeping two implementations is a real cost, accepted because the alternative is
// shipping the rule in production code. Drift is self-detecting: if this twin
// stops matching the Python one, the byte comparison it exists to enable fails.

const PATH_ROW_MARKERS = ["✓ ", "○ ", "- ", "• ", "✕ ", "✎ ", "↻ ", "  "] as const;
const PATH_ROW_TOKEN = "<ABSOLUTE PATH>";
const ABSOLUTE_PATH_RE = /\/[^\s,'"]+/g;
const MIN_STANDALONE_FRAGMENT = 24;
// A wrapped path's final chunk can carry the punctuation that followed it in the
// sentence, because wrapping splits on whitespace rather than on tokens.
const TRAILING_PUNCTUATION = /[,;:.]+$/;
// `cardPrefixed` indents an item's continuation rows by two spaces, which is the
// root-independent signal that a row belongs to the item above it.
const CONTINUATION_MARKER = "  ";

/** Python `absolute_paths_in`: longest first, so a parent cannot corrupt a child. */
export function absolutePathsIn(message: string): string[] {
  const found = new Set<string>();
  for (const match of message.matchAll(ABSOLUTE_PATH_RE)) {
    found.add(match[0].replace(/\.+$/, ""));
  }
  return [...found].sort((a, b) => b.length - a.length);
}

/** Python `_card_row_content`. */
function cardRowContent(line: string): { fragment: string; marker: string } {
  if (!(line.startsWith("│ ") && line.endsWith("│"))) return { fragment: "", marker: "" };
  const inner = line.slice(2, -1).replace(/\s+$/u, "");
  for (const marker of PATH_ROW_MARKERS) {
    if (inner.startsWith(marker)) {
      return { fragment: inner.slice(marker.length).trim(), marker };
    }
  }
  return { fragment: inner, marker: "" };
}

/**
 * Python `normalize_path_rows`.
 *
 * Only a `/`-leading row may OPEN a run. Without that, a relativized row
 * (`docs/project/…`) is also a substring of the absolute path and the rule would
 * collapse the `artifacts_materialized` rows too, erasing content that is
 * machine-independent and must stay graded.
 *
 * A continuation row may carry trailing punctuation, because wrapping splits on
 * whitespace and `<path>,` is one word.
 *
 * A new run opens on the card's list MARKER, never on a leading `/`: where a
 * wrapped path splits depends on the prefix's LENGTH, so opening at every
 * `/`-leading row makes the row count depend on the root. That is not theory — the
 * same four paths normalized to eight rows under the generator's repo-relative root
 * and four under this test's `/tmp` root, which is how the flaw was found.
 */
export function normalizePathRows(text: string, absolutePaths: readonly string[]): string {
  if (absolutePaths.length === 0) return text;
  const normalized: string[] = [];
  let runPaths: string[] | null = null;
  for (const line of text.split("\n")) {
    const { fragment, marker } = cardRowContent(line);
    if (!fragment) {
      runPaths = null;
      normalized.push(line);
      continue;
    }
    const candidate = fragment.replace(TRAILING_PUNCTUATION, "");
    const matches = candidate ? absolutePaths.filter((path) => path.includes(candidate)) : [];
    let opensAllowed = true;
    if (marker !== "" && marker !== CONTINUATION_MARKER) {
      runPaths = null;
    } else if (runPaths !== null && matches.some((path) => runPaths?.includes(path))) {
      continue;
    } else {
      opensAllowed = runPaths === null;
    }
    if (
      opensAllowed &&
      fragment.startsWith("/") &&
      matches.length > 0 &&
      candidate.length >= MIN_STANDALONE_FRAGMENT
    ) {
      normalized.push(`${`│ ${marker}${PATH_ROW_TOKEN}`.padEnd(57)}│`);
      runPaths = matches;
      continue;
    }
    runPaths = null;
    normalized.push(line);
  }
  return normalized.join("\n");
}

/** Python `project_relative`. */
export function projectRelative(path: string, projectRoot: string): string {
  const relative = path.startsWith(projectRoot) ? path.slice(projectRoot.length) : null;
  if (relative === null) return "<OUTSIDE PROJECT>";
  const trimmed = relative.replace(/^\/+/, "");
  return trimmed === "" ? "." : trimmed;
}

/** Python `scrub_message`, scoped to one project root. */
export function scrubMessage(message: string, projectRoot: string): string {
  let result = message;
  for (const absolute of absolutePathsIn(message)) {
    if (!absolute.startsWith(projectRoot)) continue;
    result = result.replaceAll(absolute, projectRelative(absolute, projectRoot));
  }
  return result;
}
